### DISCUSSION TITLE
CP outputs

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#26349] [Q&A Modules: Solid mechanics]
Hello, 

Working with Compute MultipleCrystalPlasticityStress and CrystalPlasticityStressUpdateBase, I see how I can get the slip increments (\Delta \gamma) as an output, but is there a way to get the cumulated slip rate or the plastic velocity gradient (Lp) directly or should I assign a new attribute with these values? thank you.

**Comment by GiudGiud** [comment]
Hello

You can use an auxkernel with a dedicated variable for accumulating quandtities. Every time step you sum to the accumulation variableParsedAux will do it. I think there is also a SumAux but it might be only in the thermal hydraulics module. There might an accumulateAux not sure

Guillaume

**Reply by sapitts** [reply]
Turns out `SumAux` is only available in the thermal hydraulics module. Any chance we could pull that out to framework? The capability would be nice to have in these tensor mechanics simulations too; hopefully that's enough of a generalized-usage argument to support moving the code to framework

**Reply by GiudGiud** [reply]
Tensor mechanics actually created an AccumulateAux so you can use that there.
It would be nice to avoid recoding +, -, % in each module separately haha

**Comment by sapitts** [comment]
Hi @amassaf,
I quite like @GiudGiud's recommendations for the accumulating quantities for the slip increment (note: these are calculated in the code as a rate-dependent quantity, so if you want the increment, you'll have to multiply by the timestep size for each time step). 

The plastic velocity gradient is available as a `RankTwoTensor` with the material property name `plastic_deformation_gradient` (or `base_name_plastic_deformation_gradient` if you are supplying a `base_name` argument in the input file). You can either output specific components of the tensor with `RankTwoAux` as the auxkernel, or as a scalar quantity that makes sense for your problem with `RankTwoScalarAux`.

Hope this helps!

**Reply by ghost** [reply]
Hi,
Thank you for this info I will so as suggested for the slip increments. But just to be sure, are you saying that the plastic_deformation_gradient actually returns the plastic velocity gradient (Lp) and not the usual (Fp) with Lp = \dot{Fp}Fp^{-1}?

**Reply by sapitts** [reply]
Good catch, and please accept my apologies for having given you the incorrect information for the tensor you are interested in.

We sum up the contributions from the slip increments, to calculate the (actual) plasticity velocity gradient in this method in the crystal plasticity constitutive model base class:
https://github.com/idaholab/moose/blob/3b9776e604835f35c90e594729960391e0fe93ee/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C#L461-L468
As you can see, the method does not assign the calculated tensor to any material property that can be accessed with an auxkernel.

We are happy to take pull requests, either in the base class above or in the calling class, https://github.com/idaholab/moose/blob/3b9776e604835f35c90e594729960391e0fe93ee/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L507 if you are planning to implement a code change to access this tensor

**Reply by ghost** [reply]
Thank you so much this is very helpful. I will look into it.

**Comment by ngrilli** [comment]
@amassaf  an example of exporting and visualizing slip_increment here:

https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/SlipGradients/polycrystal_initial_GND.i

**Reply by ghost** [reply]
Hi, thank you but I'm actually interested in the slip rate total and not increment
